### ISSUE TITLE
Get rid of the "flipping behaviour" in narrowign by equality/identity

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#21365]
We have [this code](https://github.com/python/mypy/blob/3a0207b6796e82ca61826d9adb4b7c95a2ff50e1/mypy/checkexpr.py#L6494-L6499) that makes sure that if the intersection of declared and narrowed is not representable, then we prefer the narrowed. I added this logic a while ago, since this is what a user usually expects in case of e.g. an assignment or even an `isinstance()` check.

However, the recent effort to make equality/identity narrowing symmetric also caused this to result in a "flipping" behavior, when sides exchange types after a comparison, see e.g. `testNarrowGenericCallableEquality`. Although such cases are rare, I think the resulting behavior is really weird/confusing.

So I propose to add a flag to each restriction in binder (similar to `from_assignment`), say `prefer_declared` that would be set to `True` for right-most operand in equality/identity. This flag would override the fallback mentioned above. So that after something like `assert x == y == a`, the type of `a` will not change if the intersection is not representable.

I understand this breaks the symmetry, but this is purely practical thing, this is why people write `x == 2` and not `2 == x`.

cc @hauntsaninja

**Comment by ilevkivskyi** [comment]
Also to be clear, if people agree here, I volunteer to do the work.

**Comment by hauntsaninja** [comment]
Sounds worth trying to me! I don't have a strong philosophical stance here. I'm fine with the symmetry break, I also think the symmetric flipping doesn't seem bad in practice. Or I guess maybe my actual philosophical stance here is that in the long run we will have more intersection types :-)

**Comment by svenpanne** [comment]
Is this issue about the following bug?

```python
from typing import reveal_type

def foo[A, B](a: A, b: B) -> None:
    if a == b:
        reveal_type((a, b))  # Revealed type is "tuple[B, A]"
    else:
        reveal_type((a, b))  # Revealed type is "tuple[A, B]"
```

If not, what is the corresponding issue? We need various suppressions in our codebase because of that, and both pyright and pyrefly infer `tuple[A, B]` for both branches, as expected.
